File: Good Debugging.cpp
#include <bits/stdc++.h>
using namespace std;

class LazySegTree {
private:
    vector<int> seg, lazy;
    int size;
    int f(int x, int y) { return x + y; }
    int default_val = 0;

public:
    LazySegTree(int N) {
        int layers = ceil(log2(N));
        size = 1 << layers;
        seg.assign(2 * size, default_val);
        lazy.assign(2 * size, default_val);
        for (int i = 0; i < N; i++){
            seg[i+size]= 1;
        }
        for (int i = size-1; i > 0; i--){
            seg[i]=seg[i*2]+seg[i*2+1];
        }
    }

    void push_down(int i, int segment_size) {
        if (lazy[i] % 2 == 0) return;
        lazy[2*i] += lazy[i];
        lazy[2*i + 1] += lazy[i];
        seg[2*i] = segment_size - seg[2*i];
        seg[2*i + 1] = segment_size - seg[2*i + 1];
        lazy[i] = default_val;
    }

    void update(int i, int l, int r, int cur_l, int cur_r, int diff) {
        if (cur_r < l || r < cur_l) return;
        int mid = (cur_l + cur_r) / 2;
        if (l <= cur_l && cur_r <= r) {
            lazy[i] += diff;
            seg[i] = (cur_r - cur_l + 1) - seg[i];
            return;
        }
        push_down(i, cur_r - mid);
        update(2*i, l, r, cur_l, mid, diff);
        update(2*i + 1, l, r, mid + 1, cur_r, diff);
        seg[i] = f(seg[2*i], seg[2*i + 1]);
    }

    int query(int i, int l, int r, int cur_l, int cur_r, int broken) {
        if (seg[i] < broken) return -1;
        if (cur_l == cur_r) return cur_l;
        int mid = (cur_l + cur_r) / 2;
        push_down(i, cur_r - mid);
        int left = query(2*i, l, r, cur_l, mid, broken);
        if (left != -1) return left;
        return query(2*i + 1, l, r, mid + 1, cur_r, broken - seg[2*i]);
    }

    int query_range(int l, int r, int broken) {
        return query(1, l, r, 0, size - 1, broken);
    }

    void update_range(int l, int r, int diff) {
        update(1, l, r, 0, size - 1, diff);
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M, L;
    cin >> N >> M >> L;

    LazySegTree seg(N);

    for (int i = 0; i < M; i++) {
        int l, r;
        cin >> l >> r;
        seg.update_range(l - 1, r - 1, 1);
        int broke = seg.query_range(0, N - 1, L);
        if (broke == -1) {
            cout << "AC?" << "\n";
        } else {
            cout << broke + 1 << "\n";
        }
    }

    return 0;
}

File: Good Debugging.py
# TLE, python is too slow, check c++ code
# https://dmoj.ca/problem/goode
# similar to https://dmoj.ca/problem/bsfast but uses lazy propagation
# use lazy segment tree to update range, traverse tree to get index of L-th broken line

import sys
from math import log2, ceil


class LazySegTree:
    def __init__(self, arr, f=lambda x, y: x + y, default=0) -> None:
        """create 1-indexed segment tree from 0-indexed array"""
        self.f = f  # function used to combine segments
        self.default = default
        layers = ceil(log2(N))  # max depth of the seg tree, root has depth 0
        self.size = 1 << layers
        self.seg = [default] * (2 * self.size)  # 1-indexed, need an extra layer for the actual data
        self.lazy = [default] * (2 * self.size)

        for i in range(N):  # base layer
            self.seg[self.size + i] = arr[i]
        for i in reversed(range(1, self.size)):  # create other layers from base
            self.seg[i] = f(self.seg[i * 2], self.seg[i * 2 + 1])

    def push_down(self, i, segment_size):
        """transfer lazy tag from parent to children"""
        if self.lazy[i] & 1 == 0:  # check if even because 2 updates cancel out
            return
        self.lazy[i * 2] += self.lazy[i]  # transfer lazy tags
        self.lazy[i * 2 + 1] += self.lazy[i]
        self.seg[i * 2] = segment_size - self.seg[i * 2]  # update value of segments
        self.seg[i * 2 + 1] = segment_size - self.seg[i * 2 + 1]
        self.lazy[i] = self.default

    def update(self, i, l, r, cur_l, cur_r, diff) -> None:
        """add diff to [left, right]"""
        if cur_r < l or r < cur_l:  # fully outside segment
            return
        mid = (cur_l + cur_r) // 2
        if l <= cur_l and cur_r <= r:  # fully inside segment: update lazily
            self.lazy[i] += diff
            self.seg[i] = (cur_r - cur_l + 1) - self.seg[i]
            return
        self.push_down(i, cur_r - mid)  # partial covers, recurse deeper
        self.update(i * 2, l, r, cur_l, mid, diff)
        self.update(i * 2 + 1, l, r, mid + 1, cur_r, diff)
        self.seg[i] = self.f(self.seg[i * 2], self.seg[i * 2 + 1])

    def query(self, i, l, r, cur_l, cur_r, broken) -> int:
        """query [left, right]"""
        if self.seg[i] < broken:  # fully outside segment or didn't break yet
            return -1
        if cur_l == cur_r:  # reached leaf
            return cur_l

        mid = (cur_l + cur_r) // 2  # traverse down further
        self.push_down(i, cur_r - mid)
        left = self.query(i * 2, l, r, cur_l, mid, broken)
        if left != -1:  # go right, minus te broken ones in the left
            return left
        return self.query(i * 2 + 1, l, r, mid + 1, cur_r, broken - self.seg[i * 2])

    def query_range(self, l, r, broken):
        return self.query(1, l, r, 0, self.size - 1, broken)

    def update_range(self, l, r, diff):
        return self.update(1, l, r, 0, self.size - 1, diff)


input = sys.stdin.readline
N, M, L = map(int, input().split())
arr = [1] * N

seg = LazySegTree(arr)

for _ in range(M):
    l, r = map(int, input().split())
    seg.update_range(l - 1, r - 1, 1)  # 0-indexed updates
    broke = seg.query_range(0, N - 1, L)  # 0-indexed queries
    if broke == -1:
        print("AC?")
    else:
        print(broke + 1)

"""
4 2 1
1 4
1 3

AC?
1
"""
